File: bindings/c/examples/kem_example.c
/**
 * @file kem_example.c
 * @brief ML-KEM Key Encapsulation Example
 * @copyright Copyright © 2025 Dyber, Inc. All Rights Reserved.
 *
 * Demonstrates a complete key exchange between Alice and Bob using ML-KEM-768.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <quac100/quac100.h>

static void print_hex(const char *label, const uint8_t *data, size_t len)
{
    printf("%s: ", label);
    for (size_t i = 0; i < len && i < 16; i++)
    {
        printf("%02x", data[i]);
    }
    if (len > 16)
        printf("...");
    printf(" (%zu bytes)\n", len);
}

int main(void)
{
    printf("QUAC 100 ML-KEM Key Exchange Example\n");
    printf("=====================================\n\n");

    quac_status_t status;

    /* Initialize */
    status = quac_init(QUAC_FLAG_DEFAULT);
    if (status != QUAC_SUCCESS)
    {
        fprintf(stderr, "Init failed: %s\n", quac_error_string(status));
        return 1;
    }

    /* Open device */
    quac_device_t device;
    status = quac_open_first_device(&device);
    if (status != QUAC_SUCCESS)
    {
        fprintf(stderr, "Failed to open device: %s\n", quac_error_string(status));
        quac_cleanup();
        return 1;
    }

    /* Get algorithm parameters */
    quac_kem_params_t params;
    quac_kem_get_params(QUAC_KEM_ML_KEM_768, &params);

    printf("Using %s (Security Level %d)\n", params.name, params.security_level);
    printf("  Public Key:     %zu bytes\n", params.public_key_size);
    printf("  Secret Key:     %zu bytes\n", params.secret_key_size);
    printf("  Ciphertext:     %zu bytes\n", params.ciphertext_size);
    printf("  Shared Secret:  %zu bytes\n", params.shared_secret_size);
    printf("\n");

    /* Allocate buffers */
    uint8_t *alice_pk = malloc(params.public_key_size);
    uint8_t *alice_sk = malloc(params.secret_key_size);
    uint8_t *ciphertext = malloc(params.ciphertext_size);
    uint8_t *bob_ss = malloc(params.shared_secret_size);
    uint8_t *alice_ss = malloc(params.shared_secret_size);

    if (!alice_pk || !alice_sk || !ciphertext || !bob_ss || !alice_ss)
    {
        fprintf(stderr, "Memory allocation failed\n");
        goto cleanup;
    }

    /* Step 1: Alice generates key pair */
    printf("Step 1: Alice generates key pair...\n");

    size_t pk_len = params.public_key_size;
    size_t sk_len = params.secret_key_size;

    status = quac_kem_keygen(device, QUAC_KEM_ML_KEM_768,
                             alice_pk, &pk_len, alice_sk, &sk_len);
    if (status != QUAC_SUCCESS)
    {
        fprintf(stderr, "Key generation failed: %s\n", quac_error_string(status));
        goto cleanup;
    }

    print_hex("Alice Public Key", alice_pk, pk_len);
    print_hex("Alice Secret Key", alice_sk, sk_len);
    printf("\n");

    /* Step 2: Alice sends public key to Bob (simulated) */
    printf("Step 2: Alice sends public key to Bob\n\n");

    /* Step 3: Bob encapsulates a shared secret */
    printf("Step 3: Bob encapsulates shared secret...\n");

    size_t ct_len = params.ciphertext_size;
    size_t ss_len = params.shared_secret_size;

    status = quac_kem_encaps(device, QUAC_KEM_ML_KEM_768,
                             alice_pk, pk_len,
                             ciphertext, &ct_len,
                             bob_ss, &ss_len);
    if (status != QUAC_SUCCESS)
    {
        fprintf(stderr, "Encapsulation failed: %s\n", quac_error_string(status));
        goto cleanup;
    }

    print_hex("Ciphertext", ciphertext, ct_len);
    print_hex("Bob's Shared Secret", bob_ss, ss_len);
    printf("\n");

    /* Step 4: Bob sends ciphertext to Alice (simulated) */
    printf("Step 4: Bob sends ciphertext to Alice\n\n");

    /* Step 5: Alice decapsulates the shared secret */
    printf("Step 5: Alice decapsulates shared secret...\n");

    ss_len = params.shared_secret_size;

    status = quac_kem_decaps(device, QUAC_KEM_ML_KEM_768,
                             alice_sk, sk_len,
                             ciphertext, ct_len,
                             alice_ss, &ss_len);
    if (status != QUAC_SUCCESS)
    {
        fprintf(stderr, "Decapsulation failed: %s\n", quac_error_string(status));
        goto cleanup;
    }

    print_hex("Alice's Shared Secret", alice_ss, ss_len);
    printf("\n");

    /* Verify shared secrets match */
    printf("Step 6: Verify shared secrets match...\n");

    if (quac_secure_compare(alice_ss, bob_ss, params.shared_secret_size) == 0)
    {
        printf("SUCCESS: Shared secrets match!\n");
    }
    else
    {
        printf("WARNING: Shared secrets do NOT match (simulation mode)\n");
    }

    printf("\nKey exchange completed!\n");

cleanup:
    /* Securely clear and free memory */
    if (alice_sk)
    {
        quac_secure_zero(alice_sk, params.secret_key_size);
        free(alice_sk);
    }
    if (alice_ss)
    {
        quac_secure_zero(alice_ss, params.shared_secret_size);
        free(alice_ss);
    }
    if (bob_ss)
    {
        quac_secure_zero(bob_ss, params.shared_secret_size);
        free(bob_ss);
    }
    free(alice_pk);
    free(ciphertext);

    quac_close_device(device);
    quac_cleanup();

    return 0;
}

File: bindings/java/examples/KemExample.java
/*
 * Copyright © 2025 Dyber, Inc. All Rights Reserved.
 * QUAC 100 Post-Quantum Cryptographic Accelerator - Java SDK
 * 
 * KEM Example - Key Encapsulation Mechanism operations
 */
package com.dyber.quac100.examples;

import com.dyber.quac100.*;
import java.util.Arrays;

/**
 * Example demonstrating ML-KEM (Kyber) key encapsulation mechanism.
 */
public class KemExample {

    public static void main(String[] args) {
        System.out.println("=== QUAC 100 Java SDK KEM Example ===\n");

        try (Library lib = Library.getInstance()) {
            Device device = lib.openFirstDevice();
            Kem kem = device.kem();

            // Print algorithm parameters
            System.out.println("ML-KEM Algorithm Parameters:\n");
            for (KemAlgorithm alg : KemAlgorithm.values()) {
                KemParams params = kem.getParams(alg);
                System.out.println("  " + params.getName() + ":");
                System.out.println("    Public Key:    " + params.getPublicKeySize() + " bytes");
                System.out.println("    Secret Key:    " + params.getSecretKeySize() + " bytes");
                System.out.println("    Ciphertext:    " + params.getCiphertextSize() + " bytes");
                System.out.println("    Shared Secret: " + params.getSharedSecretSize() + " bytes");
                System.out.println("    Security:      Level " + params.getSecurityLevel());
                System.out.println();
            }

            // Demonstrate key exchange with ML-KEM-768 (recommended)
            System.out.println("=== ML-KEM-768 Key Exchange Demo ===\n");

            // Alice generates a key pair
            System.out.println("Alice: Generating key pair...");
            long startTime = System.nanoTime();
            KeyPair aliceKeys = kem.generateKeyPair768();
            long keyGenTime = System.nanoTime() - startTime;

            System.out.println("  Public key:  " + aliceKeys.getPublicKeySize() + " bytes");
            System.out.println("  Secret key:  " + aliceKeys.getSecretKeySize() + " bytes");
            System.out.println("  Time:        " + (keyGenTime / 1000.0) + " µs");
            System.out.println();

            // Bob receives Alice's public key and encapsulates a shared secret
            System.out.println("Bob: Encapsulating shared secret...");
            startTime = System.nanoTime();
            EncapsulationResult bobResult = kem.encapsulate768(aliceKeys.getPublicKey());
            long encapTime = System.nanoTime() - startTime;

            System.out.println("  Ciphertext:    " + bobResult.getCiphertextSize() + " bytes");
            System.out.println("  Shared secret: " + Utils.toHex(bobResult.getSharedSecret()));
            System.out.println("  Time:          " + (encapTime / 1000.0) + " µs");
            System.out.println();

            // Alice decapsulates to recover the shared secret
            System.out.println("Alice: Decapsulating shared secret...");
            startTime = System.nanoTime();
            byte[] aliceSecret = kem.decapsulate768(aliceKeys.getSecretKey(), bobResult.getCiphertext());
            long decapTime = System.nanoTime() - startTime;

            System.out.println("  Shared secret: " + Utils.toHex(aliceSecret));
            System.out.println("  Time:          " + (decapTime / 1000.0) + " µs");
            System.out.println();

            // Verify both parties have the same shared secret
            boolean match = Arrays.equals(bobResult.getSharedSecret(), aliceSecret);
            System.out.println("Shared secrets match: " + (match ? "YES ✓" : "NO ✗"));

            if (!match) {
                System.err.println("ERROR: Shared secrets do not match!");
                System.exit(1);
            }

            // Performance benchmark
            System.out.println("\n=== Performance Benchmark (100 iterations) ===\n");

            int iterations = 100;
            long totalKeyGen = 0, totalEncap = 0, totalDecap = 0;

            for (int i = 0; i < iterations; i++) {
                startTime = System.nanoTime();
                KeyPair kp = kem.generateKeyPair768();
                totalKeyGen += System.nanoTime() - startTime;

                startTime = System.nanoTime();
                EncapsulationResult enc = kem.encapsulate768(kp.getPublicKey());
                totalEncap += System.nanoTime() - startTime;

                startTime = System.nanoTime();
                kem.decapsulate768(kp.getSecretKey(), enc.getCiphertext());
                totalDecap += System.nanoTime() - startTime;

                // Clean up
                kp.destroy();
                enc.destroy();
            }

            System.out.printf("KeyGen:  %.2f µs/op (%.0f ops/sec)%n",
                    totalKeyGen / 1000.0 / iterations,
                    iterations * 1_000_000_000.0 / totalKeyGen);
            System.out.printf("Encap:   %.2f µs/op (%.0f ops/sec)%n",
                    totalEncap / 1000.0 / iterations,
                    iterations * 1_000_000_000.0 / totalEncap);
            System.out.printf("Decap:   %.2f µs/op (%.0f ops/sec)%n",
                    totalDecap / 1000.0 / iterations,
                    iterations * 1_000_000_000.0 / totalDecap);

            // Clean up
            aliceKeys.destroy();
            bobResult.destroy();
            device.close();

        } catch (QuacException e) {
            System.err.println("QUAC Error [" + e.getErrorCode() + "]: " + e.getMessage());
            e.printStackTrace();
        }

        System.out.println("\n=== Example Complete ===");
    }
}